### ISSUE TITLE
@index does not properly handle explicit null values

### ISSUE THREAD

**Issue by cjihrig** [aws-amplify/amplify-cli#9915]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### How did you install the Amplify CLI?

Not applicable

### If applicable, what version of Node.js are you using?

Not applicable

### Amplify CLI Version

7.6.22

### What operating system are you using?

Not applicable

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No

### Amplify Categories

api

### Amplify Commands

Not applicable

### Describe the bug

Given the schema provided in this issue, the following query passes:

```graphql
mutation MyMutation {
  createTestModel(input: {content: "Test content", id: "902cd218-799e-4b66-9d24-20a0d7406d24"}) {
    content
    createdAt
    id
    parentId
    updatedAt
  }
}
```

But, this query fails:

```graphql
mutation MyMutation {
  createTestModel(input: {content: "Test content", id: "902cd218-799e-4b66-9d24-20a0d7406d24", parentId: null}) {
    content
    createdAt
    id
    parentId
    updatedAt
  }
}
```

Notice that `parentId` is omitted in the first query, but explicitly set to `null` in the second query. The error observed is:

```
{
  "data": {
    "createTestModel": null
  },
  "errors": [
    {
      "path": [
        "createTestModel"
      ],
      "data": null,
      "errorType": "DynamoDB:DynamoDbException",
      "errorInfo": null,
      "locations": [
        {
          "line": 2,
          "column": 3,
          "sourceName": null
        }
      ],
      "message": "One or more parameter values were invalid: Type mismatch for Index Key parentId Expected: S Actual: NULL IndexName: parent-id-index (Service: DynamoDb, Status Code: 400, Request ID: 5QJUK5D1TV31CJTREGEKQ10EOVVV4KQNSO5AEMVJF66Q9ASUAAJG)"
    }
  ]
}
```

The problematic query generates the following template:

```js
{
  "version": "2018-05-29",
  "operation": "PutItem",
  "attributeValues": {
    "createdAt": {
      "S": "2022-03-04T20:24:14.031Z"
    },
    "__typename": {
      "S": "TestModel"
    },
    "id": {
      "S": "902cd218-799e-4b66-9d24-20a0d7406d24"
    },
    "content": {
      "S": "Test content"
    },
    "parentId": {
      "NULL": null
    },
    "updatedAt": {
      "S": "2022-03-04T20:24:14.031Z"
    }
  },
  "condition": {
    "expression": "(attribute_not_exists(#id))",
    "expressionNames": {
      "#id": "id"
    }
  },
  "key": {
    "id": {
      "S": "902cd218-799e-4b66-9d24-20a0d7406d24"
    }
  }
}
```

The only lines that are different from the successful query are:

```
"parentId": {	
  "NULL": null	
},
```

### Expected behavior

Both queries behave similarly.

### Reproduction steps

1. Push the GraphQL schema listed in this issue.
2. Execute the query specified in this issue in the AppSync console.

### GraphQL schema(s)

<details>

```graphql
type TestModel @model {
  id: ID!
  content: String!
  parentId: ID @index(name: "parent-id-index")
}
```

</details>

### Log output

Not applicable

### Additional information

This bug reproduces with a vanilla GraphQL API. DataStore is not needed to reproduce the bug.

**Comment by shri-onecup** [comment]
@lazpavel this issue still happens as of CLI 8.0.1

I tried using Android models generated by Amplify codegen, to try and push "null" value as well as unspecified value for an index and the push never synced with the cloud. On CloudWatch, I can see the following error:

```
"fieldInError": true,
    "errors": [
        "CustomTemplateException(message=One or more parameter values were invalid: Type mismatch for Index Key random_id Expected: S Actual: NULL IndexName: byRandom (Service: DynamoDb, Status Code: 400, Request ID: PMUBED03DU0D4MJLT4L4ODTAJBVV4KQNSO5AEMVJF66Q9ASUAAJG), errorType=DynamoDB:DynamoDbException, data=null, errorInfo=null)"
```

**Comment by undefobj** [comment]
@lazpavel This issue seems to still exist even with the PR above merged. Can you investigate?

**Comment by lazpavel** [comment]
Hi @shri-onecup, the PR was released with CLI 8.0.2, can you confirm if you still have the issue after upgrading, note that for the fix to work the resolvers need to be regenerated (you can force that by adding an empty space in your schema)

**Comment by shri-onecup** [comment]
> Hi @shri-onecup, the PR was released with CLI 8.0.2, can you confirm if you still have the issue after upgrading, note that for the fix to work the resolvers need to be regenerated (you can force that by adding an empty space in your schema)

@lazpavel I switched to CLI 8.0.2 and it still did not work.
With your solution, I ran `amplify api gql-compile` and forced my resolvers to regenerate and it started working afterwards.

In conclusion, the bug seems to be fixed AFTER regenerating resolvers.

Thanks for the fix!
